### DISCUSSION TITLE
Is it possible to mock static methods?

### DISCUSSION THREAD

**Discussion by SKProCH** [connorivy/MockMe#19] [Q&A]
If you are using harmony/il code patching seems like mocking a static methods should be possible, isn't it?

**Comment by connorivy** [comment]
Definitely possible in theory, but not currently possible in MockMe. I'm open to implementing it. Do you have any idea what kind of api you would want for that functionality?

**Reply by SKProCH** [reply]
Probably something like a expression, e.g.:
```
MockMe.Static(() => SomeStatic.Method())
```
But i've don't quite familiar with this, so i think you can use whatever you like.

**Reply by marchewek** [reply]
How about having it a bit more consistent?
```csharp
var mock = Mock.Me<MyType>();
mock.Setup.InstanceMethod()...;
mock.TypeSetup.StaticMethod()...;
```
to put them into an example:
```csharp
var dateTimeMock = Mock.Me<DateTime>();
dateTimeMock.TypeSetup.Now.Returns(someConstantDate);
```
or `StaticSetup` for that matter. Personally I prefer `Type` as it's the scope of static methods.

**Reply by SKProCH** [reply]
This is doesn't work with fully static classes, as they cannot be used as type arguments, but overall looks good.

**Reply by marchewek** [reply]
Good point.
When a class if fully static, then I'd suggest the `Setup` node to either be empty or dropped at all.
Analogously, when a class would contain no static members, the `TypeSetup` node would either be empty or dropped at all.
Out of the two alternatives, I lean towards the latter, however it's probably more complex (if at all possible).

**Reply by marchewek** [reply]
Or maybe, because of different scopes, the following API would be better:
```csharp
var typeMock = Mock.Type<T>();
typeMock.StaticMethod()...

var mock = Mock.Instance<T>();
mock.InstanceMethod()...

ConsumeT(mock); // implicit conversion operator
```
